keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1416 @@
+5. Kubernetes scheduler
+
+The first two are driver specific. The sidecar will be provided by
+Kubernetes-CSI, but how it is used is determined when deploying the
+CSI driver. Steps 3 to 5 are explained below.
+
+### Gathering capacity information
+
+A sidecar, external-provisioner in this proposal, will be extended to
+handle the management of the new objects. This follows the normal
+approach that integration into Kubernetes is managed as part of the
+CSI driver deployment, ideally without having to modify the CSI driver
+itself. This will work without changes for some storage systems while
+others may have to support [a new CSI API
+call](#with-central-controller-generic-solution).
+
+### Pod scheduling
+
+The Kubernetes scheduler watches the capacity information and excludes
+nodes with insufficient remaining capacity when it comes to making
+scheduling decisions for a pod which uses persistent volumes with
+delayed binding. If the driver does not indicate that it supports
+capacity reporting, then the scheduler proceeds just as it does now,
+so nothing changes for existing CSI driver deployments.
+
+## Design Details
+
+### API
+
+#### CSIStorageCapacity
+
+```
+// CSIStorageCapacity stores the result of one CSI GetCapacity call for one
+// driver, one topology segment, and the parameters of one storage class.
+type CSIStorageCapacity struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+	// there are no conflicts with other CSI drivers on the cluster, the recommendation
+	// is to use csisc-<uuid>.
+	//
+	// Objects are not namespaced.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+    // Spec contains the properties that describe one capacity value.
+    Spec CSIStorageCapacitySpec
+}
+
+// CSIStorageCapacitySpec contains the properties that describe one capacity value.
+type CSIStorageCapacitySpec struct {
+	// The CSI driver that provides the storage.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
+
+    // NodeTopology defines which nodes have access to the storage for which
+    // capacity was reported. If not set, the storage is accessible from all
+    // nodes in the cluster.
+	// +optional
+	NodeTopology *v1.NodeSelector
+
+	// The storage class name of the StorageClass which provided the
+    // additional parameters for the GetCapacity call.
+	StorageClassName string
+
+	// Capacity is the value reported by the CSI driver in its GetCapacityResponse.
+    // Depending on how the driver is implemented, this might be the total
+    // size of the available storage which is only available when allocating
+    // multiple smaller volumes ("total available capacity") or the
+    // actual size that a volume may have ("maximum volume size").
```

Comment:
+1

Reply:
That would send a signal to CSI driver developers that they have to provide "total capacity" in their `GetCapacityResponse` although that will work less well for the initial purpose (comparing against volume sizes) than "maximum volume size".

If we want to resolve ambiguity at the API level from the start, then we need to add both. I would prefer that myself, but @msau42 wanted the API to match what we get from CSI today.

Reply:
Actually, @msau42 made me realize that is not just a problem for autoscaler, but also for the scheduler.

To maximize throughput (and avoid starving pods that might not be using volumes), volume binding happens asynchronously from scheduling decisions. For this reason, if the scheduler can't simulate (to some degree) how capacity is affected with a pod, it might schedule a second pod in a node that will no longer have the capacity. If that is the case, the volume binding for the 2nd pod will eventually fail and the pod will be added back to the queue, wasting scheduling cycles.

I acknowledge that we won't be able to perfectly model the consumption of the capacity, but having (perhaps) optional fields such as totalCapacity would be necessary to have low scheduling latency.

And my preference would be that this is changed in CSI too, if you can help us drive that change.

Reply:
I want to avoid deviating from CSI as much as possible unless the CSI project says that this is a use case they don't want to support. 

So regarding total available vs max volume size, when I look at the original CSI PR that added the capacity field, the examples  given seem to imply the intention was "total available size": https://github.com/container-storage-interface/spec/pull/85, so I think we should start with that. I agree that this doesn't work well for storage systems that are prone to fragmentation, but I think we should try to resolve it through the CSI spec first before putting it into the Kubernetes API.

As for the topic of "total available" vs "total capacity (used + available)", will add my thoughts in the other comment thread.

Reply:
I have created https://github.com/container-storage-interface/spec/issues/432 to get some clarity into the CSI spec.